src/transform/src/predicate_propagation.rs

```diff
@@ -0,0 +1,576 @@
+                output_knowledge
+            }
+            MirRelationExpr::Join {
+                inputs,
+                equivalences,
+                ..
+            } => {
+                let mut knowledge = Vec::new();
+
+                // Collect input knowledge, but update column references.
+                let mut prior_arity = 0;
+                for input in inputs.iter_mut() {
+                    for mut predicate in PredicateKnowledge::action(input, let_knowledge)? {
+                        predicate.visit_mut(&mut |expr| {
+                            if let MirScalarExpr::Column(c) = expr {
+                                *c += prior_arity;
+                            }
+                        });
+                        knowledge.push(predicate);
+                    }
+                    prior_arity += input.arity();
+                }
+
+                let structured = PredicateStructure::new(&knowledge);
+                let mut new_predicates = Vec::new();
+
+                for equivalence in equivalences.iter_mut() {
+                    for expr in equivalence.iter_mut() {
+                        optimize(expr, &structured);
+                    }
+
+                    if let Some(first) = equivalence.get(0) {
+                        for other in equivalence[1..].iter() {
+                            new_predicates
+                                .push(first.clone().call_binary(other.clone(), BinaryFunc::Eq));
+                        }
+                    }
+                }
+
+                knowledge.extend(new_predicates);
+                normalize_predicates(&mut knowledge);
+                knowledge
+            }
+            MirRelationExpr::Reduce {
+                input,
+                group_key,
+                aggregates,
+                ..
+            } => {
+                let input_type = input.typ();
+                let input_knowledge = PredicateKnowledge::action(input, let_knowledge)?;
+                let structured = PredicateStructure::new(&input_knowledge);
+                for key in group_key.iter_mut() {
+                    optimize(key, &structured);
+                }
+                for aggregate in aggregates.iter_mut() {
+                    optimize(&mut aggregate.expr, &structured);
+                }
+
+                // List of predicates we will return.
+                let mut predicates = Vec::new();
+
+                // Predicates that depend only on group keys should remain true of those columns.
+                // 0. Form a map from key expressions to the columns they will become.
+                let mut key_exprs = HashMap::new();
+                for (index, expr) in group_key.iter().enumerate() {
+                    key_exprs.insert(expr.clone(), MirScalarExpr::Column(index));
+                }
+                // 1. Visit each predicate, and collect those that reach no `ScalarExpr::Column` when substituting per `key_exprs`.
```

Comment:
Isn't it the other way around? I mean, we collect those that only reference columns in the grouping key.